src/fldmod_by_const.jl

```diff
@@ -0,0 +1,147 @@
+"""
+    calculate_inv_coeff(::Type{T}, f)
+
+Returns the magic numbers needed for `fldmod_by_const(x, Val(C))`: `2^N / C`, or the
+`inv_coeff`. That `inv_coeff` has all leading zeros shifted away to maximize precision, so
+this also returns how many bits need to be shifted back to undo that optimization, making
+the return value: `((2^N / C) << toshift, toshift)`.
+"""
+# This function is marked `@pure` because it has no side-effects and depends on no global
+# state. The value will never change for a given (type,precision) pair.
+# This allows its result to be const-folded away when called with Const values.
+Base.@pure function calculate_inv_coeff(::Type{T}, f) where {T}
+    # First, calculate 2^nbits(T)/f
+    # We calculate it at double-precision and shift away leading zeros to preserve the most
+    # precision. Later, we will shift the final answer back to undo this operation.
+    UT = unsigned(T)
+    invcoeff = twoToTheSizeOf(UT)÷f
+    toshift = _leading_zeros(invcoeff)
+    invcoeff = invcoeff << toshift
+    # Now, truncate the result to sizeof(T), rounding to maintain precision.
+    invcoeff = _round_to_even(fldmod(invcoeff, typemax(UT))..., typemax(UT)) % T
+    return invcoeff, toshift
+end
+# These are needed to handle Int128, which widens to BigInt
+twoToTheSizeOf(::Type{T}) where {T} = typemax(widen(T))
+twoToTheSizeOf(::Union{Type{Int128}, Type{UInt128}}) = BigInt(2)^256
+twoToTheSizeOf(::Type{BigInt}) = BigInt(2)^256
+
+_leading_zeros(x) = leading_zeros(x)
+_leading_zeros(x::BigInt) = leading_zeros((x >> 128) % UInt128)
+
+
+# Splits `x` into `(hi,lo)` and returns them as the corresponding narrowed type
+function splitint(x::T) where {T<:Integer}
+    NT = narrow(T)
+    (x >> nbits(NT)) % NT, x % NT
+end
+
+narrow(::Type{Int128}) = Int64
+narrow(::Type{Int64}) = Int32
+narrow(::Type{Int32}) = Int16
+narrow(::Type{Int16}) = Int8
+narrow(::Type{UInt128}) = UInt64
+narrow(::Type{UInt64}) = UInt32
+narrow(::Type{UInt32}) = UInt16
+narrow(::Type{UInt16}) = UInt8
+
+"""    nbits(x) = sizeof(x)*8 """
+nbits(x) = sizeof(x)*8
+
+# Implemenation based on umul32hi, from https://stackoverflow.com/a/22847373/751061
+# /* compute the upper 32 bits of the product of two unsigned 32-bit integers */
+@inline function unsigned_splitmul_upper(a::T, b::T) where T<:Unsigned
```

Comment:
I'm fuzzy on why this is function is needed for T=Unsigned .. can you just do a 128-bit multiply and shift right 64 bits?

Reply:
Ah, sorry, here's a Julia-ism that's different than C. `Unsigned` is an abstract base class for all unsigned integers.

So this function applies to any of `UInt8, UInt16, UInt32, UInt64, UInt128`, etc. It also works for the `UInt256` that I add in the next PR, https://github.com/RelationalAI-oss/FixedPointDecimals.jl/pull/6.

Even though I tested these functions for all those types, in reality I'm only ever going to call them with `UInt128` and up, because the check in `custom_fldmod` at https://github.com/RelationalAI-oss/FixedPointDecimals.jl/pull/5/files#diff-ea4e0198fdaaab3db0fa802fdb33c161R174 will only ever call `fldmod_by_const` with `UInt128` and up. So for this PR it's just `UInt128`, but in the next PR it's also `UInt256`.